lazyflow/operators/opCompressedUserLabelArray.py

```diff
@@ -247,12 +262,24 @@ def _executeProjection2D(self, roi, destination):
             # Compute slicing within the deep array and slicing within this block
             deep_relative_intersection = numpy.subtract(intersecting_roi, input_roi.start)
             block_relative_intersection = numpy.subtract(intersecting_roi, block_start)
-                        
-            deep_data = self._getBlockDataset( entire_block_roi )[roiToSlice(*block_relative_intersection)]
+            block_relative_intersection_slicing = roiToSlice(*block_relative_intersection)
+
+            block = self._getBlockDataset( entire_block_roi )
+            deep_data = None
+            if self.Output.meta.has_mask:
+                deep_data = numpy.ma.masked_array(
+                    block["data"][block_relative_intersection_slicing],
+                    mask=block["mask"][block_relative_intersection_slicing],
+                    fill_value=block["fill_value"][()],
+                    shrink=False
+                )
+            else:
+                deep_data = block[block_relative_intersection_slicing]
+
+            # make binary and convert to float (must copy)
+            deep_data_float = deep_data.astype(numpy.float32)
+            deep_data_float[deep_data_float.nonzero()] = 1
 
-            # make binary and convert to float
-            deep_data_float = numpy.where( deep_data, numpy.float32(1.0), numpy.float32(0.0) )
-            
             # multiply by slice-index
```

Comment:
> Wait, so... are you saying that numpy.where() can't be used in lazyflow any more? Perhaps we need to define a helper function that transparently handles both cases (masked/unmasked). Or maybe even monkey-patch numpy.where()?
> 
> What other ordinary numpy functions will we need to avoid now that we want to support masked arrays?

That is a very interesting question.

If we want to generalize the use of masked arrays further in the pipeline, we need to think about what they mean in the context of other operators. At this point, I haven't tried to address this question generally as it is not clear to me for all operators what a mask should mean. Certainly, I have some thoughts on this, but am interested in hearing others. I wouldn't want to force anybody into a particular definition until we at least had some discussion.

NumPy generally uses the mask to indicate erroneous regions or regions that should be ignored for calculation. For instance, division by zero results in masking and calculating means completely exclude masked values. Matplotlib and qimage2ndarray treat the mask as transparent (i.e. ignored). In Volumina, we are now consistently following the latter practice (it was already followed to some extent before just not consistently). It’s not clear to me what practice we want to follow in Lazyflow. We could follow the NumPy practice exactly, or we could only ignore values, or we could have some other use cases in mind.

As a result, I tried to steer clear of answering this question by focusing only on a subset of cache operators, which were intended, despite some sophisticated internals, to behave as identity operators. In this case, I would expect a masked array to propagated through intact. However, I did tweak `OpCompressedUserLabelArray` to propagate the mask and this is not exactly an identity operator. Another general principle I tried to follow is that a `numpy.ndarray` should behave as a masked array with a mask of `False`.

As far as the internals are concerned, we would want to check whether to use `numpy.where` or `numpy.ma.where`. Unfortunately, `numpy.ma.where` on normal `numpy.ndarray`s results in a masked array. I agree with you that we would want to avoid having people do this themselves. So, having some helper functions in the interim is a good idea. This is why I have made changes to `lazyflow.stype` and have tried to more generally use methods from there in the operators that I changed. In the case of `numpy.where`, I think a PR is in order here and I would be happy to do it. Though `numpy.where` is an exception, not the norm. Using `numpy.asarray` is bad. It is better to use `numpy.asanyarray`. Functions from `numpy` normally work fine with masked arrays. Most of the work here was just making sure the proper array was allocated when needed.